CONTRIBUTING.md

```diff
@@ -1,174 +1,163 @@
+
+- Your title should be clear, concise, and adequately summarise the idea.
+
+- Avoid exclusively talking about specific solutions to problems - your
+feature request should describe the general case for why a feature should be
+added. Focus on who your feature request would help, when it would help them
+and why.
+
+- That said, please feel free to suggest hypothetical API designs, as long
+as they're not the focus.
+
+### We'll work with you to develop it further
+
+- Most feature requests originate from valid concerns, but might need work to
+craft an API that's suitable for Fusion.
+
+- We'll close the issue once an implementation of the design is approved and
+merged into a main branch of Fusion.
+
+### Tips for good feature requests
+
+- We typically prefer users declaring information upfront, rather than
+procedural solutions that depend on execution order.
+
+- Guide users towards writing code that's more optimal, but not at the expense
+of maintainability.
+
+- Aim to make errors impossible to represent, or at least very easy to spot.
+
+### Tips for bad feature requests
+
+- Overly specific solutions. Niche features often would be better served as
+part of third party libraries; Fusion deals with concerns that everyone has an
+interest in.
+
+- Code is read many times, but written very few times. 'Saving keystrokes' is
+not a strong enough reason to include a feature, and can actually harm
+readability and maintainability quite seriously.
+
+- It's almost always possible to express an idea with simple function and table
+primitives; simpler implementations are less buggy, easier to understand for
+maintainers and users, and often more performant.
+
+-----
+
+## I want to contribute code
+
+While we do accept community pull requests, please observe that Fusion has a
+very high bar for code and API quality. We're used by a lot of important people
+and projects!
+
+### Respect the issues
+
+- New features without existing feature requests are closed on principle. Pull
+requests aren't the place to introduce new ideas suddenly.
+
+- Please don't implement new features before they're ready to be implemented;
+*this has happened a lot.* We often end up throwing out a lot of work because
+the design was not yet approved and ended up changing direction.
+
+- Bug fixes without existing bug reports are passable, but information about
+the bug must be provided, like a regular bug report would contain.
+
+### Create a new branch and draft a pull request
+
+- Ensure it's clear what you're working on, why you're working on it and what
+you aim to achieve with the pull request.
+
+- Keep your pull requests small and specifically targeted; for example, by
+separating different features into different pull requests.
+
+- Doing this as early as possible means we can observe and comment on your
+work as you go, enabling us to help you or to point out potential
+shortcomings while it's still early on and easy to rectify.
```

Comment:
I presume this comment refers to "draft a pull request" from the header, but it's a bit "outdone" by the two comments that come first in this section. It's also not obvious what "draft a pull request" actually means, that should maybe be explained. (I assume it's referring to the draft status of PRs that GitHub offers for public repos)